In [1]:
from nselib import capital_market as ca

In [28]:
market_indices=ca.market_watch_all_indices()

In [29]:
market_indices

key                                      index  \
0   BROAD MARKET INDICES                                   NIFTY 50   
1   BROAD MARKET INDICES                              NIFTY NEXT 50   
2   BROAD MARKET INDICES                                  NIFTY 100   
3   BROAD MARKET INDICES                                  NIFTY 200   
4   BROAD MARKET INDICES                                  NIFTY 500   
..                   ...                                        ...   
72  FIXED INCOME INDICES  NIFTY 10 YR BENCHMARK G-SEC (CLEAN PRICE)   
73  FIXED INCOME INDICES                   NIFTY 4-8 YR G-SEC INDEX   
74  FIXED INCOME INDICES                 NIFTY 11-15 YR G-SEC INDEX   
75  FIXED INCOME INDICES          NIFTY 15 YR AND ABOVE G-SEC INDEX   
76  FIXED INCOME INDICES                NIFTY COMPOSITE G-SEC INDEX   

          indexSymbol      last  variation  percentChange      open      high  \
0            NIFTY 50  23398.90      75.95           0.33  23480.95  23481.05   
1       NIFTY NEXT 50  70842.55     640.40           0.91  70831.40  70933.55   
2           NIFTY 100  24449.50     100.70           0.41  24527.05  24527.15   
3           NIFTY 200  13282.50      62.55           0.47  13317.65  13317.70   
4           NIFTY 500  22074.30     113.35           0.52  22120.75  22120.80   
..                ...       ...        ...            ...       ...       ...   
72  NIFTY GS 10YR CLN    879.02       1.43           0.16    877.59    879.31   
73     NIFTY GS 4 8YR   2845.26       3.26           0.11   2842.56   2845.37   
74   NIFTY GS 11 15YR   2909.14       5.27           0.18   2904.43   2909.14   
75  NIFTY GS 15YRPLUS   3215.27       6.71           0.20   3209.19   3216.53   
76  NIFTY GS COMPSITE   2720.77       4.30           0.15   2717.00   2720.83   

         low  previousClose  ...    dy  declines advances unchanged  \
0   23353.90       23322.95  ...  1.24        15       35         0   
1   70231.80       70202.15  ...  0.97        22       28         0   
2   24376.75       24348.80  ...  1.15        38       63         0   
3   13237.85       13219.95  ...  1.08        74      127         0   
4   21992.15       21960.95  ...  1.04       186      313         3   
..       ...            ...  ...   ...       ...      ...       ...   
72    877.59         877.59  ...             NaN      NaN       NaN   
73   2842.56        2842.00  ...             NaN      NaN       NaN   
74   2904.43        2903.87  ...             NaN      NaN       NaN   
75   3209.19        3208.56  ...             NaN      NaN       NaN   
76   2717.00        2716.47  ...             NaN      NaN       NaN   

   perChange365d perChange30d previousDay oneWeekAgo oneMonthAgo  oneYearAgo  
0          24.61         5.51    23264.85   22821.40    22104.05    18716.15  
1          62.21         9.09    69678.70   68224.45    64354.40    43278.65  
2          30.43         6.07    24271.05   23806.95    22956.35    18668.00  
3          33.96         6.53    13162.70   12902.20    12410.00     9868.80  
4          36.84         7.07    21852.00   21365.00    20511.00    16048.20  
..           ...          ...         ...        ...         ...         ...  
72         -0.57         0.43      877.58     877.34      874.05      882.84  
73          6.88         0.91     2842.17    2837.68     2816.42     2659.11  
74           7.7         1.30     2903.50    2899.46     2867.13     2696.83  
75          8.81         1.52     3207.19    3200.37     3161.31     2949.69  
76           7.3         1.11     2715.96    2711.01     2687.22     2532.14  

[77 rows x 24 columns]

In [6]:
market_indices.key

0     BROAD MARKET INDICES
1     BROAD MARKET INDICES
2     BROAD MARKET INDICES
3     BROAD MARKET INDICES
4     BROAD MARKET INDICES
              ...         
72    FIXED INCOME INDICES
73    FIXED INCOME INDICES
74    FIXED INCOME INDICES
75    FIXED INCOME INDICES
76    FIXED INCOME INDICES
Name: key, Length: 77, dtype: object

In [10]:
indicess_name=market_indices.key.value_counts()

In [11]:
indicess_name

key
STRATEGY INDICES        19
BROAD MARKET INDICES    18
THEMATIC INDICES        17
SECTORAL INDICES        16
FIXED INCOME INDICES     7
Name: count, dtype: int64

In [16]:
type(indicess_name.index)

pandas.core.indexes.base.Index

In [18]:
for i in indicess_name.index:
    print(i)

STRATEGY INDICES
BROAD MARKET INDICES
THEMATIC INDICES
SECTORAL INDICES
FIXED INCOME INDICES


In [20]:
indicess_name.index[0]

'STRATEGY INDICES'

In [31]:
market_indices[market_indices.key==indicess_name.index[0]]

key                            index        indexSymbol  \
34  STRATEGY INDICES  NIFTY DIVIDEND OPPORTUNITIES 50  NIFTY DIV OPPS 50   
35  STRATEGY INDICES          NIFTY GROWTH SECTORS 15  NIFTY GROWSECT 15   
36  STRATEGY INDICES              NIFTY100 QUALITY 30  NIFTY100 QUALTY30   
37  STRATEGY INDICES                 NIFTY50 VALUE 20   NIFTY50 VALUE 20   
38  STRATEGY INDICES           NIFTY50 TR 2X LEVERAGE  NIFTY50 TR 2X LEV   
39  STRATEGY INDICES           NIFTY50 PR 2X LEVERAGE  NIFTY50 PR 2X LEV   
40  STRATEGY INDICES            NIFTY50 TR 1X INVERSE  NIFTY50 TR 1X INV   
41  STRATEGY INDICES            NIFTY50 PR 1X INVERSE  NIFTY50 PR 1X INV   
42  STRATEGY INDICES          NIFTY50 DIVIDEND POINTS  NIFTY50 DIV POINT   
43  STRATEGY INDICES                   NIFTY ALPHA 50     NIFTY ALPHA 50   
44  STRATEGY INDICES             NIFTY50 EQUAL WEIGHT    NIFTY50 EQL WGT   
45  STRATEGY INDICES            NIFTY100 EQUAL WEIGHT   NIFTY100 EQL WGT   
46  STRATEGY INDICES       NIFTY100 LOW VOLATILITY 30  NIFTY100 LOWVOL30   
47  STRATEGY INDICES              NIFTY200 QUALITY 30  NIFTY200 QUALTY30   
48  STRATEGY INDICES    NIFTY ALPHA LOW-VOLATILITY 30  NIFTY ALPHALOWVOL   
49  STRATEGY INDICES             NIFTY200 MOMENTUM 30  NIFTY200MOMENTM30   
50  STRATEGY INDICES       NIFTY MIDCAP150 QUALITY 50  NIFTY M150 QLTY50   
51  STRATEGY INDICES                NIFTY200 ALPHA 30  NIFTY200 ALPHA 30   
52  STRATEGY INDICES      NIFTY MIDCAP150 MOMENTUM 50  NIFTYM150MOMNTM50   

        last  variation  percentChange      open      high       low  \
34   6211.95      25.80           0.42   6225.65   6225.70   6186.40   
35  11095.20      22.15           0.20  11127.90  11128.10  11075.25   
36   5631.20      25.40           0.45   5636.20   5636.35   5600.80   
37  12763.00      19.65           0.15  12825.55  12825.80  12734.80   
38  19080.25     120.10           0.63  19213.70  19213.70  19007.10   
39  13214.35      83.20           0.63  13306.75  13306.75  13163.70   
40    173.30      -0.50          -0.29    172.65    173.60    172.65   
41    208.00      -0.65          -0.31    207.25    208.40    207.25   
42    100.12       0.00           0.00    100.12    100.12    100.12   
43  55720.25     530.20           0.96  55661.40  55776.95  55346.60   
44  29901.15     166.60           0.56  29942.60  29943.75  29774.60   
45  32266.85     186.55           0.58  32316.70  32317.30  32097.65   
46  18854.35      15.30           0.08  18928.80  18929.80  18784.25   
47  20573.15      98.75           0.48  20587.60  20591.40  20464.70   
48  27887.25      89.20           0.32  27967.45  27967.60  27735.45   
49  35848.30     354.30           1.00  35807.70  35890.50  35445.90   
50  24249.85     259.25           1.08  24141.60  24283.95  24102.80   
51  27146.55     443.25           1.66  26961.05  27186.65  26766.25   
52  62846.30     454.50           0.73  62929.70  63146.85  62532.95   

    previousClose  ...    dy  declines advances unchanged perChange365d  \
34        6186.15  ...  2.49        18       32         0         48.08   
35       11073.05  ...  1.99         7        8         0         18.07   
36        5605.80  ...  1.66        11       19         0         32.44   
37       12743.35  ...  2.17         6       14         0         31.68   
38       18960.15  ...             NaN      NaN       NaN          46.3   
39       13131.15  ...             NaN      NaN       NaN         42.86   
40         173.80  ...             NaN      NaN       NaN        -16.58   
41         208.65  ...             NaN      NaN       NaN        -15.59   
42         100.12  ...             NaN      NaN       NaN         35.55   
43       55190.05  ...  0.64        17       33         0         81.75   
44       29734.55  ...  1.24        15       35         0         35.86   
45       32080.30  ...  1.05        38       63         0         47.43   
46       18839.05  ...   1.2        11       19         0          32.7   
47   

In [27]:
market_indices.key.value_counts()

key
STRATEGY INDICES    77
Name: count, dtype: int64

In [37]:
market_indices["index"].value_counts().sort_values(ascending=True)

index
NIFTY 50                             1
NIFTY 500                            1
NIFTY MIDCAP 50                      1
NIFTY MIDCAP 100                     1
NIFTY SMALLCAP 100                   1
                                    ..
NIFTY 15 YR AND ABOVE G-SEC INDEX    1
NIFTY MNC                            1
NIFTY MIDCAP LIQUID 15               1
NIFTY 8-13 YR G-SEC                  1
NIFTY COMPOSITE G-SEC INDEX          1
Name: count, Length: 77, dtype: int64

In [41]:
market_indices["last"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 77 entries, 0 to 76
Series name: last
Non-Null Count  Dtype  
--------------  -----  
77 non-null     float64
dtypes: float64(1)
memory usage: 744.0 bytes


In [42]:
import pandas as pd

In [44]:
market_indices.itertuples(25)

In [45]:
market_indices.interpolate()

key                                      index  \
0   BROAD MARKET INDICES                                   NIFTY 50   
1   BROAD MARKET INDICES                              NIFTY NEXT 50   
2   BROAD MARKET INDICES                                  NIFTY 100   
3   BROAD MARKET INDICES                                  NIFTY 200   
4   BROAD MARKET INDICES                                  NIFTY 500   
..                   ...                                        ...   
72  FIXED INCOME INDICES  NIFTY 10 YR BENCHMARK G-SEC (CLEAN PRICE)   
73  FIXED INCOME INDICES                   NIFTY 4-8 YR G-SEC INDEX   
74  FIXED INCOME INDICES                 NIFTY 11-15 YR G-SEC INDEX   
75  FIXED INCOME INDICES          NIFTY 15 YR AND ABOVE G-SEC INDEX   
76  FIXED INCOME INDICES                NIFTY COMPOSITE G-SEC INDEX   

          indexSymbol      last  variation  percentChange      open      high  \
0            NIFTY 50  23398.90      75.95           0.33  23480.95  23481.05   
1       NIFTY NEXT 50  70842.55     640.40           0.91  70831.40  70933.55   
2           NIFTY 100  24449.50     100.70           0.41  24527.05  24527.15   
3           NIFTY 200  13282.50      62.55           0.47  13317.65  13317.70   
4           NIFTY 500  22074.30     113.35           0.52  22120.75  22120.80   
..                ...       ...        ...            ...       ...       ...   
72  NIFTY GS 10YR CLN    879.02       1.43           0.16    877.59    879.31   
73     NIFTY GS 4 8YR   2845.26       3.26           0.11   2842.56   2845.37   
74   NIFTY GS 11 15YR   2909.14       5.27           0.18   2904.43   2909.14   
75  NIFTY GS 15YRPLUS   3215.27       6.71           0.20   3209.19   3216.53   
76  NIFTY GS COMPSITE   2720.77       4.30           0.15   2717.00   2720.83   

         low  previousClose  ...    dy  declines advances unchanged  \
0   23353.90       23322.95  ...  1.24        15       35         0   
1   70231.80       70202.15  ...  0.97        22       28         0   
2   24376.75       24348.80  ...  1.15        38       63         0   
3   13237.85       13219.95  ...  1.08        74      127         0   
4   21992.15       21960.95  ...  1.04       186      313         3   
..       ...            ...  ...   ...       ...      ...       ...   
72    877.59         877.59  ...             NaN      NaN       NaN   
73   2842.56        2842.00  ...             NaN      NaN       NaN   
74   2904.43        2903.87  ...             NaN      NaN       NaN   
75   3209.19        3208.56  ...             NaN      NaN       NaN   
76   2717.00        2716.47  ...             NaN      NaN       NaN   

   perChange365d perChange30d previousDay oneWeekAgo oneMonthAgo  oneYearAgo  
0          24.61         5.51    23264.85   22821.40    22104.05    18716.15  
1          62.21         9.09    69678.70   68224.45    64354.40    43278.65  
2          30.43         6.07    24271.05   23806.95    22956.35    18668.00  
3          33.96         6.53    13162.70   12902.20    12410.00     9868.80  
4          36.84         7.07    21852.00   21365.00    20511.00    16048.20  
..           ...          ...         ...        ...         ...         ...  
72         -0.57         0.43      877.58     877.34      874.05      882.84  
73          6.88         0.91     2842.17    2837.68     2816.42     2659.11  
74           7.7         1.30     2903.50    2899.46     2867.13     2696.83  
75          8.81         1.52     3207.19    3200.37     3161.31     2949.69  
76           7.3         1.11     2715.96    2711.01     2687.22     2532.14  

[77 rows x 24 columns]

In [46]:
market_indices

key                                      index  \
0   BROAD MARKET INDICES                                   NIFTY 50   
1   BROAD MARKET INDICES                              NIFTY NEXT 50   
2   BROAD MARKET INDICES                                  NIFTY 100   
3   BROAD MARKET INDICES                                  NIFTY 200   
4   BROAD MARKET INDICES                                  NIFTY 500   
..                   ...                                        ...   
72  FIXED INCOME INDICES  NIFTY 10 YR BENCHMARK G-SEC (CLEAN PRICE)   
73  FIXED INCOME INDICES                   NIFTY 4-8 YR G-SEC INDEX   
74  FIXED INCOME INDICES                 NIFTY 11-15 YR G-SEC INDEX   
75  FIXED INCOME INDICES          NIFTY 15 YR AND ABOVE G-SEC INDEX   
76  FIXED INCOME INDICES                NIFTY COMPOSITE G-SEC INDEX   

          indexSymbol      last  variation  percentChange      open      high  \
0            NIFTY 50  23398.90      75.95           0.33  23480.95  23481.05   
1       NIFTY NEXT 50  70842.55     640.40           0.91  70831.40  70933.55   
2           NIFTY 100  24449.50     100.70           0.41  24527.05  24527.15   
3           NIFTY 200  13282.50      62.55           0.47  13317.65  13317.70   
4           NIFTY 500  22074.30     113.35           0.52  22120.75  22120.80   
..                ...       ...        ...            ...       ...       ...   
72  NIFTY GS 10YR CLN    879.02       1.43           0.16    877.59    879.31   
73     NIFTY GS 4 8YR   2845.26       3.26           0.11   2842.56   2845.37   
74   NIFTY GS 11 15YR   2909.14       5.27           0.18   2904.43   2909.14   
75  NIFTY GS 15YRPLUS   3215.27       6.71           0.20   3209.19   3216.53   
76  NIFTY GS COMPSITE   2720.77       4.30           0.15   2717.00   2720.83   

         low  previousClose  ...    dy  declines advances unchanged  \
0   23353.90       23322.95  ...  1.24        15       35         0   
1   70231.80       70202.15  ...  0.97        22       28         0   
2   24376.75       24348.80  ...  1.15        38       63         0   
3   13237.85       13219.95  ...  1.08        74      127         0   
4   21992.15       21960.95  ...  1.04       186      313         3   
..       ...            ...  ...   ...       ...      ...       ...   
72    877.59         877.59  ...             NaN      NaN       NaN   
73   2842.56        2842.00  ...             NaN      NaN       NaN   
74   2904.43        2903.87  ...             NaN      NaN       NaN   
75   3209.19        3208.56  ...             NaN      NaN       NaN   
76   2717.00        2716.47  ...             NaN      NaN       NaN   

   perChange365d perChange30d previousDay oneWeekAgo oneMonthAgo  oneYearAgo  
0          24.61         5.51    23264.85   22821.40    22104.05    18716.15  
1          62.21         9.09    69678.70   68224.45    64354.40    43278.65  
2          30.43         6.07    24271.05   23806.95    22956.35    18668.00  
3          33.96         6.53    13162.70   12902.20    12410.00     9868.80  
4          36.84         7.07    21852.00   21365.00    20511.00    16048.20  
..           ...          ...         ...        ...         ...         ...  
72         -0.57         0.43      877.58     877.34      874.05      882.84  
73          6.88         0.91     2842.17    2837.68     2816.42     2659.11  
74           7.7         1.30     2903.50    2899.46     2867.13     2696.83  
75          8.81         1.52     3207.19    3200.37     3161.31     2949.69  
76           7.3         1.11     2715.96    2711.01     2687.22     2532.14  

[77 rows x 24 columns]

In [47]:
ca.price_volume_and_deliverable_position_data()

TypeError: price_volume_and_deliverable_position_data() missing 1 required positional argument: 'symbol'

In [50]:
testdata1=ca.bulk_deal_data("17-05-2024","13-06-2024")

In [51]:
testdata1

Date      Symbol              SecurityName  \
0     17-MAY-2024        AGUL     A G Universal Limited   
1     17-MAY-2024        AGUL     A G Universal Limited   
2     17-MAY-2024  BLUEPEBBLE       Blue Pebble Limited   
3     17-MAY-2024  BLUEPEBBLE       Blue Pebble Limited   
4     17-MAY-2024    CROMPTON  Crompt Grea Con Elec Ltd   
...           ...         ...                       ...   
1762  13-JUN-2024    VENUSREM    Venus Remedies Limited   
1763  13-JUN-2024   VISHWARAJ   Vishwaraj Sugar Ind Ltd   
1764  13-JUN-2024   VISHWARAJ   Vishwaraj Sugar Ind Ltd   
1765  13-JUN-2024  WILLAMAGOR     Williamson Magor & Co   
1766  13-JUN-2024  WILLAMAGOR     Williamson Magor & Co   

                                ClientName Buy/Sell QuantityTraded  \
0                     SUHAS RAJU VASISHTHA      BUY          2,000   
1                     SUHAS RAJU VASISHTHA     SELL         58,000   
2                      NIKUNJ KAUSHIK SHAH      BUY         60,000   
3                      NIKUNJ KAUSHIK SHAH     SELL         60,000   
4            GRAVITON RESEARCH CAPITAL LLP      BUY      32,30,372   
...                                    ...      ...            ...   
1762         GRAVITON RESEARCH CAPITAL LLP     SELL         68,957   
1763                  HRTI PRIVATE LIMITED      BUY       8,84,345   
1764                  HRTI PRIVATE LIMITED     SELL       9,47,205   
1765  SILVER LINE VENTURES PRIVATE LIMITED      BUY         75,000   
1766  SILVER LINE VENTURES PRIVATE LIMITED     SELL          4,861   

     TradePrice/Wght.Avg.Price Remarks  
0                        51.45       -  
1                        51.58       -  
2                       289.24       -  
3                       280.97       -  
4                       387.01       -  
...                        ...     ...  
1762                    377.21       -  
1763                     17.19       -  
1764                     17.26       -  
1765                     41.39       -  
1766                     39.85       -  

[1767 rows x 8 columns]

In [53]:
testdata1.columns

Index(['Date', 'Symbol', 'SecurityName', 'ClientName', 'Buy/Sell',
       'QuantityTraded', 'TradePrice/Wght.Avg.Price', 'Remarks'],
      dtype='object')

In [55]:
testdata1.head()

Date      Symbol              SecurityName  \
0  17-MAY-2024        AGUL     A G Universal Limited   
1  17-MAY-2024        AGUL     A G Universal Limited   
2  17-MAY-2024  BLUEPEBBLE       Blue Pebble Limited   
3  17-MAY-2024  BLUEPEBBLE       Blue Pebble Limited   
4  17-MAY-2024    CROMPTON  Crompt Grea Con Elec Ltd   

                      ClientName Buy/Sell QuantityTraded  \
0           SUHAS RAJU VASISHTHA      BUY          2,000   
1           SUHAS RAJU VASISHTHA     SELL         58,000   
2            NIKUNJ KAUSHIK SHAH      BUY         60,000   
3            NIKUNJ KAUSHIK SHAH     SELL         60,000   
4  GRAVITON RESEARCH CAPITAL LLP      BUY      32,30,372   

  TradePrice/Wght.Avg.Price Remarks  
0                     51.45       -  
1                     51.58       -  
2                    289.24       -  
3                    280.97       -  
4                    387.01       -

In [56]:
testdata1.tail()

Date      Symbol             SecurityName  \
1762  13-JUN-2024    VENUSREM   Venus Remedies Limited   
1763  13-JUN-2024   VISHWARAJ  Vishwaraj Sugar Ind Ltd   
1764  13-JUN-2024   VISHWARAJ  Vishwaraj Sugar Ind Ltd   
1765  13-JUN-2024  WILLAMAGOR    Williamson Magor & Co   
1766  13-JUN-2024  WILLAMAGOR    Williamson Magor & Co   

                                ClientName Buy/Sell QuantityTraded  \
1762         GRAVITON RESEARCH CAPITAL LLP     SELL         68,957   
1763                  HRTI PRIVATE LIMITED      BUY       8,84,345   
1764                  HRTI PRIVATE LIMITED     SELL       9,47,205   
1765  SILVER LINE VENTURES PRIVATE LIMITED      BUY         75,000   
1766  SILVER LINE VENTURES PRIVATE LIMITED     SELL          4,861   

     TradePrice/Wght.Avg.Price Remarks  
1762                    377.21       -  
1763                     17.19       -  
1764                     17.26       -  
1765                     41.39       -  
1766                     39.85       -

In [57]:
testdata1.columns

Index(['Date', 'Symbol', 'SecurityName', 'ClientName', 'Buy/Sell',
       'QuantityTraded', 'TradePrice/Wght.Avg.Price', 'Remarks'],
      dtype='object')

In [63]:
testdata1.dtypes

Date                         object
Symbol                       object
SecurityName                 object
ClientName                   object
Buy/Sell                     object
QuantityTraded               object
TradePrice/Wght.Avg.Price    object
Remarks                      object
dtype: object

In [64]:
testdata1.describe()

Date     Symbol               SecurityName  \
count          1767       1767                       1767   
unique           20        331                        330   
top     11-JUN-2024  SANGINITA  Sanginita Chemicals Limit   
freq            310         92                         92   

                           ClientName Buy/Sell QuantityTraded  \
count                            1767     1767           1767   
unique                            440        2           1058   
top     GRAVITON RESEARCH CAPITAL LLP     SELL         70,000   
freq                              338      918             17   

       TradePrice/Wght.Avg.Price Remarks  
count                       1767    1767  
unique                      1284       1  
top                        16.60       -  
freq                          40    1767

In [65]:
testdata1.columns

Index(['Date', 'Symbol', 'SecurityName', 'ClientName', 'Buy/Sell',
       'QuantityTraded', 'TradePrice/Wght.Avg.Price', 'Remarks'],
      dtype='object')

In [76]:
testdata1["QuantityTraded"]=testdata1["QuantityTraded"].str.replace(",","").astype("int")

In [78]:
testdata1.dtypes

Date                         object
Symbol                       object
SecurityName                 object
ClientName                   object
Buy/Sell                     object
QuantityTraded                int32
TradePrice/Wght.Avg.Price    object
Remarks                      object
dtype: object

In [79]:
testdata1

Date      Symbol              SecurityName  \
0     17-MAY-2024        AGUL     A G Universal Limited   
1     17-MAY-2024        AGUL     A G Universal Limited   
2     17-MAY-2024  BLUEPEBBLE       Blue Pebble Limited   
3     17-MAY-2024  BLUEPEBBLE       Blue Pebble Limited   
4     17-MAY-2024    CROMPTON  Crompt Grea Con Elec Ltd   
...           ...         ...                       ...   
1762  13-JUN-2024    VENUSREM    Venus Remedies Limited   
1763  13-JUN-2024   VISHWARAJ   Vishwaraj Sugar Ind Ltd   
1764  13-JUN-2024   VISHWARAJ   Vishwaraj Sugar Ind Ltd   
1765  13-JUN-2024  WILLAMAGOR     Williamson Magor & Co   
1766  13-JUN-2024  WILLAMAGOR     Williamson Magor & Co   

                                ClientName Buy/Sell  QuantityTraded  \
0                     SUHAS RAJU VASISHTHA      BUY            2000   
1                     SUHAS RAJU VASISHTHA     SELL           58000   
2                      NIKUNJ KAUSHIK SHAH      BUY           60000   
3                      NIKUNJ KAUSHIK SHAH     SELL           60000   
4            GRAVITON RESEARCH CAPITAL LLP      BUY         3230372   
...                                    ...      ...             ...   
1762         GRAVITON RESEARCH CAPITAL LLP     SELL           68957   
1763                  HRTI PRIVATE LIMITED      BUY          884345   
1764                  HRTI PRIVATE LIMITED     SELL          947205   
1765  SILVER LINE VENTURES PRIVATE LIMITED      BUY           75000   
1766  SILVER LINE VENTURES PRIVATE LIMITED     SELL            4861   

     TradePrice/Wght.Avg.Price Remarks  
0                        51.45       -  
1                        51.58       -  
2                       289.24       -  
3                       280.97       -  
4                       387.01       -  
...                        ...     ...  
1762                    377.21       -  
1763                     17.19       -  
1764                     17.26       -  
1765                     41.39       -  
1766                     39.85       -  

[1767 rows x 8 columns]

In [83]:
testdata1["TradePrice/Wght.Avg.Price"]=testdata1["TradePrice/Wght.Avg.Price"].str.replace(",","").astype("float")

In [84]:
testdata1.dtypes

Date                          object
Symbol                        object
SecurityName                  object
ClientName                    object
Buy/Sell                      object
QuantityTraded                 int32
TradePrice/Wght.Avg.Price    float64
Remarks                       object
dtype: object

In [87]:
testdata1["Remarks"].value_counts()

Remarks
-    1767
Name: count, dtype: int64

In [88]:
testdata1.drop("Remarks",axis=1,inplace=True)

In [92]:
countdata=testdata1["ClientName"].value_counts()

In [97]:
countdata.index.shape

(440,)

In [101]:
countdata

ClientName
GRAVITON RESEARCH CAPITAL LLP                         338
HRTI PRIVATE LIMITED                                  110
MULTIPLIER SHARE & STOCK ADVISORS PRIVATE LIMITED      80
QE SECURITIES LLP                                      54
MANSI SHARE AND STOCK ADVISORS PVT LTD                 39
                                                     ... 
MANJULABEN BABULAL MUNGARA                              1
360 ONE ASSET MANAGEMENT LIMITED                        1
CRAFT EMERGING MARKET FUND PCC- ELITE CAPITAL FUND      1
BENGAL FIN. & INV. PVT. LTD                             1
POWER MECH INFRA PRIVATE LIMITED                        1
Name: count, Length: 440, dtype: int64

In [106]:
testdata1[testdata1["ClientName"]=="GRAVITON RESEARCH CAPITAL LLP"].sort_index(ascending=False)

Date      Symbol               SecurityName  \
1762  13-JUN-2024    VENUSREM     Venus Remedies Limited   
1761  13-JUN-2024    VENUSREM     Venus Remedies Limited   
1760  13-JUN-2024  UTTAMSUGAR  Uttam Sugar Mills Limited   
1759  13-JUN-2024  UTTAMSUGAR  Uttam Sugar Mills Limited   
1758  13-JUN-2024         USK   Udayshivakumar Infra Ltd   
...           ...         ...                        ...   
16    17-MAY-2024        MOIL               MOIL Limited   
9     17-MAY-2024        GRSE          Garden Reach Ship   
8     17-MAY-2024        GRSE          Garden Reach Ship   
5     17-MAY-2024    CROMPTON   Crompt Grea Con Elec Ltd   
4     17-MAY-2024    CROMPTON   Crompt Grea Con Elec Ltd   

                         ClientName Buy/Sell  QuantityTraded  \
1762  GRAVITON RESEARCH CAPITAL LLP     SELL           68957   
1761  GRAVITON RESEARCH CAPITAL LLP      BUY           68957   
1760  GRAVITON RESEARCH CAPITAL LLP     SELL          200884   
1759  GRAVITON RESEARCH CAPITAL LLP      BUY          200884   
1758  GRAVITON RESEARCH CAPITAL LLP     SELL          310596   
...                             ...      ...             ...   
16    GRAVITON RESEARCH CAPITAL LLP      BUY         1421230   
9     GRAVITON RESEARCH CAPITAL LLP     SELL          578862   
8     GRAVITON RESEARCH CAPITAL LLP      BUY          578862   
5     GRAVITON RESEARCH CAPITAL LLP     SELL         3206972   
4     GRAVITON RESEARCH CAPITAL LLP      BUY         3230372   

      TradePrice/Wght.Avg.Price  
1762                     377.21  
1761                     376.56  
1760                     368.52  
1759                     368.19  
1758                      71.43  
...                         ...  
16                       479.01  
9                       1052.28  
8                       1052.07  
5                        387.55  
4                        387.01  

[338 rows x 7 columns]

In [109]:
testdata1["SecurityName"].value_counts()

SecurityName
Sanginita Chemicals Limit    92
Hoac Foods India Limited     51
Paras Def and Spce Tech L    50
TruCap Finance Limited       49
Sakuma Exports Limited       37
                             ..
Aatmaj Healthcare Limited     1
Jash Engineering Limited      1
Kamdhenu Ispat Limited        1
Poonawalla Fincorp Ltd        1
Indraprastha Gas Limited      1
Name: count, Length: 330, dtype: int64

In [110]:
testdata1

Date      Symbol              SecurityName  \
0     17-MAY-2024        AGUL     A G Universal Limited   
1     17-MAY-2024        AGUL     A G Universal Limited   
2     17-MAY-2024  BLUEPEBBLE       Blue Pebble Limited   
3     17-MAY-2024  BLUEPEBBLE       Blue Pebble Limited   
4     17-MAY-2024    CROMPTON  Crompt Grea Con Elec Ltd   
...           ...         ...                       ...   
1762  13-JUN-2024    VENUSREM    Venus Remedies Limited   
1763  13-JUN-2024   VISHWARAJ   Vishwaraj Sugar Ind Ltd   
1764  13-JUN-2024   VISHWARAJ   Vishwaraj Sugar Ind Ltd   
1765  13-JUN-2024  WILLAMAGOR     Williamson Magor & Co   
1766  13-JUN-2024  WILLAMAGOR     Williamson Magor & Co   

                                ClientName Buy/Sell  QuantityTraded  \
0                     SUHAS RAJU VASISHTHA      BUY            2000   
1                     SUHAS RAJU VASISHTHA     SELL           58000   
2                      NIKUNJ KAUSHIK SHAH      BUY           60000   
3                      NIKUNJ KAUSHIK SHAH     SELL           60000   
4            GRAVITON RESEARCH CAPITAL LLP      BUY         3230372   
...                                    ...      ...             ...   
1762         GRAVITON RESEARCH CAPITAL LLP     SELL           68957   
1763                  HRTI PRIVATE LIMITED      BUY          884345   
1764                  HRTI PRIVATE LIMITED     SELL          947205   
1765  SILVER LINE VENTURES PRIVATE LIMITED      BUY           75000   
1766  SILVER LINE VENTURES PRIVATE LIMITED     SELL            4861   

      TradePrice/Wght.Avg.Price  
0                         51.45  
1                         51.58  
2                        289.24  
3                        280.97  
4                        387.01  
...                         ...  
1762                     377.21  
1763                      17.19  
1764                      17.26  
1765                      41.39  
1766                      39.85  

[1767 rows x 7 columns]

1. in the data 17 may 2024 to 13 jun 2024 available

In [111]:
testdata1

Date      Symbol              SecurityName  \
0     17-MAY-2024        AGUL     A G Universal Limited   
1     17-MAY-2024        AGUL     A G Universal Limited   
2     17-MAY-2024  BLUEPEBBLE       Blue Pebble Limited   
3     17-MAY-2024  BLUEPEBBLE       Blue Pebble Limited   
4     17-MAY-2024    CROMPTON  Crompt Grea Con Elec Ltd   
...           ...         ...                       ...   
1762  13-JUN-2024    VENUSREM    Venus Remedies Limited   
1763  13-JUN-2024   VISHWARAJ   Vishwaraj Sugar Ind Ltd   
1764  13-JUN-2024   VISHWARAJ   Vishwaraj Sugar Ind Ltd   
1765  13-JUN-2024  WILLAMAGOR     Williamson Magor & Co   
1766  13-JUN-2024  WILLAMAGOR     Williamson Magor & Co   

                                ClientName Buy/Sell  QuantityTraded  \
0                     SUHAS RAJU VASISHTHA      BUY            2000   
1                     SUHAS RAJU VASISHTHA     SELL           58000   
2                      NIKUNJ KAUSHIK SHAH      BUY           60000   
3                      NIKUNJ KAUSHIK SHAH     SELL           60000   
4            GRAVITON RESEARCH CAPITAL LLP      BUY         3230372   
...                                    ...      ...             ...   
1762         GRAVITON RESEARCH CAPITAL LLP     SELL           68957   
1763                  HRTI PRIVATE LIMITED      BUY          884345   
1764                  HRTI PRIVATE LIMITED     SELL          947205   
1765  SILVER LINE VENTURES PRIVATE LIMITED      BUY           75000   
1766  SILVER LINE VENTURES PRIVATE LIMITED     SELL            4861   

      TradePrice/Wght.Avg.Price  
0                         51.45  
1                         51.58  
2                        289.24  
3                        280.97  
4                        387.01  
...                         ...  
1762                     377.21  
1763                      17.19  
1764                      17.26  
1765                      41.39  
1766                      39.85  

[1767 rows x 7 columns]

2. in the data 1767 row and 7 column are available

In [112]:
testdata1.shape

(1767, 7)

In [113]:
testdata1.columns

Index(['Date', 'Symbol', 'SecurityName', 'ClientName', 'Buy/Sell',
       'QuantityTraded', 'TradePrice/Wght.Avg.Price'],
      dtype='object')

3. in the data column are mention above
4. the data "Quantity Trade" and "tradePrice" are integer and float values respectively

In [114]:
testdata1.count()

Date                         1767
Symbol                       1767
SecurityName                 1767
ClientName                   1767
Buy/Sell                     1767
QuantityTraded               1767
TradePrice/Wght.Avg.Price    1767
dtype: int64

5. in the data no null value are available

In [115]:
testdata1

Date      Symbol              SecurityName  \
0     17-MAY-2024        AGUL     A G Universal Limited   
1     17-MAY-2024        AGUL     A G Universal Limited   
2     17-MAY-2024  BLUEPEBBLE       Blue Pebble Limited   
3     17-MAY-2024  BLUEPEBBLE       Blue Pebble Limited   
4     17-MAY-2024    CROMPTON  Crompt Grea Con Elec Ltd   
...           ...         ...                       ...   
1762  13-JUN-2024    VENUSREM    Venus Remedies Limited   
1763  13-JUN-2024   VISHWARAJ   Vishwaraj Sugar Ind Ltd   
1764  13-JUN-2024   VISHWARAJ   Vishwaraj Sugar Ind Ltd   
1765  13-JUN-2024  WILLAMAGOR     Williamson Magor & Co   
1766  13-JUN-2024  WILLAMAGOR     Williamson Magor & Co   

                                ClientName Buy/Sell  QuantityTraded  \
0                     SUHAS RAJU VASISHTHA      BUY            2000   
1                     SUHAS RAJU VASISHTHA     SELL           58000   
2                      NIKUNJ KAUSHIK SHAH      BUY           60000   
3                      NIKUNJ KAUSHIK SHAH     SELL           60000   
4            GRAVITON RESEARCH CAPITAL LLP      BUY         3230372   
...                                    ...      ...             ...   
1762         GRAVITON RESEARCH CAPITAL LLP     SELL           68957   
1763                  HRTI PRIVATE LIMITED      BUY          884345   
1764                  HRTI PRIVATE LIMITED     SELL          947205   
1765  SILVER LINE VENTURES PRIVATE LIMITED      BUY           75000   
1766  SILVER LINE VENTURES PRIVATE LIMITED     SELL            4861   

      TradePrice/Wght.Avg.Price  
0                         51.45  
1                         51.58  
2                        289.24  
3                        280.97  
4                        387.01  
...                         ...  
1762                     377.21  
1763                      17.19  
1764                      17.26  
1765                      41.39  
1766                      39.85  

[1767 rows x 7 columns]

In [117]:
testdata1["ClientName"].unique()

array(['SUHAS RAJU VASISHTHA', 'NIKUNJ KAUSHIK SHAH',
       'GRAVITON RESEARCH CAPITAL LLP', 'ELUHIM LIMITED', 'NITN  KAPOOR',
       'ANKITA VISHAL SHAH', 'SETU SECURITIES PVT LTD', 'MOVING LIMITED',
       'DARSHAN PRADIP DOSHI',
       'ZINNIA GLOBAL FUND PCC - CELL DEWCAP FUND',
       'NEIGHBOURHOOD INVESTMENT PRIVATE LIMITED',
       'NISHCHAYA TRADINGS PRIVATE LIMITED', 'VT CAPITAL MARKET PVT LTD',
       'ALOK BANSAL', 'YASHISH DAHIYA', 'ANIL VISHANJI DEDHIA',
       'GREEN PEAKS ENTERPRISES LLP', 'PRAVEEN KUMAR  GUPTA',
       'YUGA STOCKS AND COMMODITIES PRIVATE LIMITED  .',
       'VISHAL BIPINKUMAR DOSHI', 'DINESHSINH BHIMSINH CHAVADA',
       'MULTIPLIER SHARE & STOCK ADVISORS PRIVATE LIMITED',
       'SANGINITA INDUSTRIES LLP', 'SRESTHA FINVEST LIMITED',
       'JINESH ASHWIN MATALIA', 'NALIN RAMANIKLAL SHAH',
       'BHAVNA ABHIRAM BHAI NATHWANI',
       'MANSI SHARE AND STOCK ADVISORS PVT LTD',
       'VISCO TRADE ASSOCIATES LIMITED', 'SIGMA INSIGHTS PRIVATE LIMITED',


6. in the data only 440 unique "ClientName" are available

In [118]:
testdata1

Date      Symbol              SecurityName  \
0     17-MAY-2024        AGUL     A G Universal Limited   
1     17-MAY-2024        AGUL     A G Universal Limited   
2     17-MAY-2024  BLUEPEBBLE       Blue Pebble Limited   
3     17-MAY-2024  BLUEPEBBLE       Blue Pebble Limited   
4     17-MAY-2024    CROMPTON  Crompt Grea Con Elec Ltd   
...           ...         ...                       ...   
1762  13-JUN-2024    VENUSREM    Venus Remedies Limited   
1763  13-JUN-2024   VISHWARAJ   Vishwaraj Sugar Ind Ltd   
1764  13-JUN-2024   VISHWARAJ   Vishwaraj Sugar Ind Ltd   
1765  13-JUN-2024  WILLAMAGOR     Williamson Magor & Co   
1766  13-JUN-2024  WILLAMAGOR     Williamson Magor & Co   

                                ClientName Buy/Sell  QuantityTraded  \
0                     SUHAS RAJU VASISHTHA      BUY            2000   
1                     SUHAS RAJU VASISHTHA     SELL           58000   
2                      NIKUNJ KAUSHIK SHAH      BUY           60000   
3                      NIKUNJ KAUSHIK SHAH     SELL           60000   
4            GRAVITON RESEARCH CAPITAL LLP      BUY         3230372   
...                                    ...      ...             ...   
1762         GRAVITON RESEARCH CAPITAL LLP     SELL           68957   
1763                  HRTI PRIVATE LIMITED      BUY          884345   
1764                  HRTI PRIVATE LIMITED     SELL          947205   
1765  SILVER LINE VENTURES PRIVATE LIMITED      BUY           75000   
1766  SILVER LINE VENTURES PRIVATE LIMITED     SELL            4861   

      TradePrice/Wght.Avg.Price  
0                         51.45  
1                         51.58  
2                        289.24  
3                        280.97  
4                        387.01  
...                         ...  
1762                     377.21  
1763                      17.19  
1764                      17.26  
1765                      41.39  
1766                      39.85  

[1767 rows x 7 columns]

7. In the data Buy sell exist

In [143]:
testdata1["Buy/Sell"].value_counts()

Buy/Sell
SELL    918
BUY     849
Name: count, dtype: int64

8. in the data buy and sell data difference is

In [138]:
testdata1.groupby("SecurityName")["ClientName"].count().sort_values(ascending=False).values.mean()

5.3545454545454545

In [139]:
testdata1.groupby("SecurityName")["ClientName"].count().sort_values(ascending=False).values.max()

92

In [140]:
testdata1.groupby("SecurityName")["ClientName"].count().sort_values(ascending=False).values.min()

1

In [142]:
testdata1.describe()

QuantityTraded  TradePrice/Wght.Avg.Price
count    1.767000e+03                1767.000000
mean     2.909745e+06                 334.422694
std      1.772797e+07                 602.392544
min      3.000000e+02                   0.300000
25%      1.000000e+05                  31.000000
50%      3.500000e+05                 131.000000
75%      1.051716e+06                 386.000000
max      3.971399e+08                9192.210000